lib/rubocop/cop/lint/return_in_void_context.rb

```diff
@@ -32,12 +32,17 @@ module Lint
       class ReturnInVoidContext < Base
         MSG = 'Do not return a value in `%<method>s`.'
 
+        # Returning out of these methods only exits the block itself.
+        SCOPE_CHANGING_METHODS = %i[lambda define_method define_singleton_method].freeze
+
         def on_return(return_node)
           return unless return_node.descendants.any?
 
           def_node = return_node.each_ancestor(:def).first
```

Comment:
The patch corresponds to the spec I commented on below.

```suggestion
          def_node = return_node.each_ancestor(:def, defs).first
```

Reply:
Good catch, I fixed it and added the test